src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,276 @@
+  }
+
+  private static KeyStore initializeKeyStore(String pemCaCertificate)
+          throws CertificateException
+  {
+    try {
+      PEMReader       reader      = new PEMReader(new InputStreamReader(new ByteArrayInputStream(pemCaCertificate.getBytes())));
+      X509Certificate certificate = (X509Certificate) reader.readObject();
+
+      if (certificate == null) {
+        throw new CertificateException("No certificate found in parsing!");
+      }
+
+      KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      keyStore.load(null);
+      keyStore.setCertificateEntry("ca", certificate);
+      return keyStore;
+    } catch (IOException | KeyStoreException ex) {
+      throw new CertificateException(ex.toString());
+    } catch (NoSuchAlgorithmException ex) {
+      throw new AssertionError(ex);
+    }
+  }
+
+  @Override
+  public void start() {
+    this.directoryReconciliationWorker = new DirectoryReconciliationWorker(serverApiUrl, jedisPool, accountsManager, client);
+
+    this.directoryReconciliationWorker.start();
+  }
+
+  @Override
+  public void stop() {
+    directoryReconciliationWorker.shutdown();
+  }
+
+  private static class DirectoryReconciliationWorker extends Thread {
+
+    private final String              serverApiUrl;
+    private final ReplicatedJedisPool jedisPool;
+    private final AccountsManager     accountsManager;
+    private final Client              client;
+
+    private final String workerId;
+    private final Random random;
+
+    private boolean running  = true;
+    private boolean finished = false;
+
+    DirectoryReconciliationWorker(String serverApiUrl,
+                                  ReplicatedJedisPool jedisPool,
+                                  AccountsManager accountsManager,
+                                  Client client)
+    {
+      super(DirectoryReconciliationWorker.class.getSimpleName());
+
+      this.serverApiUrl = serverApiUrl;
+      this.jedisPool = jedisPool;
+      this.accountsManager = accountsManager;
+      this.client = client;
+
+      SecureRandom secureRandom  = new SecureRandom();
+      byte[]       workerIdBytes = new byte[16];
+      secureRandom.nextBytes(new byte[16]);
+      this.workerId = Hex.toString(workerIdBytes);
+
+      this.random = new Random(secureRandom.nextLong());
+    }
+
+    synchronized void shutdown() {
+      running = false;
+      while (!finished) {
+        Util.wait(this);
+      }
+    }
+
+    synchronized boolean sleepWhileRunning(long delay) {
+      long start   = System.currentTimeMillis();
+      long elapsed = 0;
+      while (running && elapsed < delay) {
+        try {
+          wait(delay - elapsed);
+        } catch (InterruptedException ex) {
+        }
+        elapsed = System.currentTimeMillis() - start;
+      }
+      return running;
+    }
+
+    private List<String> readBucket(long bucket) {
+      Timer.Context timer = readBucketTimer.time();
+      try {
```

Comment:
you can use try with resources here iirc?

Reply:
done in 4d5b40bb442c40bcda99aaa1748b6581875479b8